scapy/contrib/isotp.py

```diff
@@ -623,12 +641,38 @@ def recv(self, x=0xffff):
 
     @staticmethod
     def select(sockets, remain=None):
-        """This function is called during sendrecv() routine to select
-        the available sockets.
+        """This function is called during sendrecv() routine to wait for
+        sockets to be ready to receive
         """
-        # ISOTPSoftSockets aren't selectable, so we return all of them
-        # sockets, None (means use the socket's recv() )
-        return sockets, None
+        blocking = remain is None or remain > 0
+
+        def find_ready_sockets():
+            return list(filter(lambda x: not x.ins.rx_queue.empty(), sockets))
+
+        ready_sockets = find_ready_sockets()
+        if len(ready_sockets) > 0 or not blocking:
+            return ready_sockets, None
+
+        exit_select = Event()
+
+        def my_cb(msg):
+            exit_select.set()
+
+        try:
+            for s in sockets:
+                s.ins.rx_callbacks.append(my_cb)
+
+            exit_select.wait(remain)
+
+        finally:
+            for s in sockets:
+                try:
+                    s.ins.rx_callbacks.remove(my_cb)
+                except ValueError:
+                    pass
+
+        ready_sockets = find_ready_sockets()
+        return ready_sockets, None
```

Comment:
We already have a similar thing in scapy/automaton.py that waits for callbacks.

There is some doc in the headers, do you mind having a look into it?

Reply:
I would not replace the current select implementation with automaton.SelectableSelector because currently it works without creating additional timeout threads for each socket (since each ISOTP socket already has a receiver thread in the background).

However, I implemented automaton.SelectableObject on the ISOTPSocketImplementation.